source/agora/crypto/Key.d

```diff
@@ -134,27 +137,30 @@ public struct PublicKey
     /// Uses Stellar's representation instead of hex
     public string toString () const @trusted nothrow
     {
-        ubyte[VersionWidth + PublicKey.sizeof + ChecksumWidth] bin;
+        scope(failure) assert(0);
+        ubyte[VersionWidth + PublicKey.sizeof] bin;
         bin[0] = VersionByte.AccountID;
-        bin[VersionWidth .. $ - ChecksumWidth] = this.data[];
-        bin[$ - ChecksumWidth .. $] = checksum(bin[0 .. $ - ChecksumWidth]);
-        return Base32.encode(bin).assumeUnique;
+        bin[VersionWidth .. $] = this.data[];
+        return encodeBech32(HumanReadablePart, bin, Encoding.Bech32m, true).
+            assumeUnique;
     }
 
     /// Make sure the sink overload of BitBlob is not picked
     public void toString (scope void delegate(const(char)[]) sink) const @trusted
     {
-        ubyte[VersionWidth + PublicKey.sizeof + ChecksumWidth] bin;
+        ubyte[VersionWidth + PublicKey.sizeof] bin;
         bin[0] = VersionByte.AccountID;
-        bin[VersionWidth .. $ - ChecksumWidth] = this.data[];
-        bin[$ - ChecksumWidth .. $] = checksum(bin[0 .. $ - ChecksumWidth]);
-        Base32.encode(bin, sink);
+        bin[VersionWidth .. $] = this.data[];
+        string encoded = encodeBech32(HumanReadablePart, bin,
+            Encoding.Bech32m, true).assumeUnique;
+        foreach (enc_char; encoded)
+            sink([enc_char]);
```

Comment:
The point of this overload is to avoid allocation, and this code completely fails this.
First and foremost, why the foreach ? It's iterating over *every element of the array* then allocating a *1-element array* to pass to the sink. Why not feed `encoded` directly to the sink ???

`encodeBech32` could be made to not allocate, but we will leave this for another PR. But I really can't understand what's the point of this `foreach`...

Reply:
This is because I'm not familiar with the concept of the `sink`. I will look into the use of the `sink` and modify it right now.